src/VSCodeExtension/src/languageServer.ts

```diff
@@ -0,0 +1,518 @@
+        );
+        this.context.subscriptions.push(
+            // The definition of the StateChangeEvent has changed in recent versions of VS Code,
+            // so we use a dictionary from enum of different states to strings to make sure that
+            // the debug log is useful even if the enum changes.
+            client.onDidChangeState(stateChangeEvent => {
+                var states : { [key in State]: string } = {
+                    [State.Running]: "running",
+                    [State.Starting]: "starting",
+                    [State.Stopped]: "stopped"
+                };
+                console.log(`[qsharp-lsp] State ${states[stateChangeEvent.oldState]} -> ${states[stateChangeEvent.newState]}`);
+            })
+        );
+
+        client.onDidChangeState(
+            (event) => {
+                if (event.oldState === State.Running && event.newState === State.Stopped) {
+                    sendTelemetryEvent(EventNames.lspStopped);
+                }
+            }
+        );
+        client.onTelemetry(forwardServerTelemetry);
+
+        // When we're ready, send the telemetry event that we started successfully.
+        client.onReady().then(() => {
+            let elapsed = Date.now() - languageServerStartedAt;
+            sendTelemetryEvent(
+                EventNames.lspReady,
+                {},
+                {
+                    'elapsedTime': elapsed
+                }
+            );
+        });
+
+        return client;
+
+    }
+
+    async start(retries : number = 3) : Promise<void> {
+        if (!await this.findExecutable()) {
+            // Try again after downloading.
+            try {
+                await this.downloadLanguageServer();
+            } catch (err) {
+                console.log(`[qsharp-lsp] Error downloading language server: ${err}. ${retries} left.`);
+                if (retries > 0) {
+                    return this.start(retries - 1);
+                } else {
+                    let retryItem = "Try again";
+                    let reportFeedbackItem = "Report feedback...";
+                    switch (await vscode.window.showErrorMessage(
+                        "Could not download Q# language server.",
+                        retryItem, reportFeedbackItem
+                    )) {
+                        case retryItem:
+                            return this.start(1);
+                            break;
+                        case reportFeedbackItem:
+                            vscode.env.openExternal(vscode.Uri.parse(
+                                "https://github.com/microsoft/qsharp-compiler/issues/new?assignees=&labels=bug,Area-IDE&template=bug_report.md&title="
+                            ));
+                            return;
+                            break;
+                    }
+                }
+            }
+            if (!this.findExecutable()) {
+                // TODO: handle this error more gracefully.
```

Comment:
As I understand that case should never happen. A more informative comment would be helpful.